Add apiserver heartbeat with conntrack

tom-hadlaw-hs opened the pull request:
:warning: :warning: **aanm note to reviewers / backporters**: This fixes a bug that needs to be backported to 1.7/1.6 but we should only do it after heavily testing it.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](http://docs.cilium.io/en/stable/contributing/contributing/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue. 
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](http://docs.cilium.io/en/stable/contributing/contributing/#dev-coo).
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Adding a periodic heartbeat to check for stale connections to the apiserver. If a request takes too long then client-go's conntrack dialer close connections function is called for all client types.

Fixes: #9792

```release-note
When running in Kubernetes, Cilium will run a periodic heartbeat and close all open Kubernetes client connections if the active connections become unresponsive. 
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10184)
<!-- Reviewable:end -->

aanm reviewed (commented):
This is a move in the right direction! Maybe we should add a cilium-agent flag that allows to user to specify the timeout and keep alive. Just to be sure, are the timeouts the same used by kubelet?

aanm commented on daemon/daemon_main.go:
```diff
@@ -712,6 +712,11 @@ func init() {
+	flags.Bool(option.EnableK8sHeartbeat, true, "Run periodic heartbeat against apiserver which resets all open connection after timeout")
+
+	flags.Duration(option.K8sHeartbeatTimeout, 30*time.Second, "Timeout for k8s heartbeat")
+
```
please put this after line 720 and add a similar line as `option.BindEnv...` for each option

aanm commented on pkg/k8s/client_test.go:
```diff
@@ -29,7 +29,7 @@ import (
 
 	. "gopkg.in/check.v1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Unnecessary changes

aanm commented on pkg/k8s/init.go:
```diff
@@ -11,7 +11,6 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
```
please re-add this new line

aanm commented on daemon/daemon_main.go:
```diff
@@ -712,6 +712,11 @@ func init() {
+
+	flags.Bool(option.EnableK8sHeartbeat, true, "Run periodic heartbeat against apiserver which resets all open connection after timeout")
+
+	flags.Duration(option.K8sHeartbeatTimeout, 30*time.Second, "Timeout for k8s heartbeat")
```
We can drop `option.EnableK8sHeartbeat` since the user can disable this feature by setting `K8sHeartbeatTimeout` with 0.

aanm commented on pkg/k8s/client.go:
```diff
@@ -105,11 +109,66 @@ func CreateConfig() (*rest.Config, error) {
+			}
+
+			close(done)
+		}()
```
I'm afraid of this. It seems harmless but that won't be the case for big clusters. `/healthz` could return a 503 and we would restart the connection, possibly make it worse.

tgraf replied:
How does kubelet cope with this?

tom-hadlaw-hs replied:
Kubelet does a get on nodes

aanm reviewed (changes requested):
@tom-hadlaw-hs this is on the right path. I'm just a little concerned with the `runHeartbeat` function. Can you try it out in your cluster without the `runHeartbeat` to see if the issue still persists? I can provision a cilium image if you make a commit reverting the `runHeartbeat` for you to try out.

aanm commented on daemon/daemon_main.go:
```diff
@@ -714,6 +714,9 @@ func init() {
 
+	flags.Duration(option.K8sHeartbeatTimeout, 30*time.Second, "Timeout for k8s heartbeat")
+	option.BindEnv(option.K8sHeartbeatTimeout)
+
```
Please also add a flag in `operator/flags.go`

tom-hadlaw-hs replied:
done

aanm commented on pkg/k8s/client.go:
```diff
@@ -105,11 +108,50 @@ func CreateConfig() (*rest.Config, error) {
+	go wait.Until(func() {
+		done := make(chan error)
+		go func() {
+			_, err := k8sCli.Interface.CoreV1().Namespaces().Get("kube-system", metav1.GetOptions{})
```
please add this comment to the code:
```
Kubernetes does a get node of the node that kubelet is running [0]. This seems excessive in our case because the amount of data transferred is bigger than doing a Get of the kube-system namespace. For this reason we have picked to perform a get on `kube-system` instead a get of a node.

[0] https://github.com/kubernetes/kubernetes/blob/v1.17.3/pkg/kubelet/kubelet_node_status.go#L423
```

aanm commented on pkg/k8s/client.go:
```diff
@@ -105,11 +108,50 @@ func CreateConfig() (*rest.Config, error) {
 
+func setDialer(config *rest.Config) func() {
+	context := (&net.Dialer{
+		Timeout:   30 * time.Second,
```
Shouldn't this be `Timeout: option.Config.K8sHeartbeatTimeout`?

tom-hadlaw-hs replied:
This is just for dial timeouts, 30s was the kubelet hardcoded default. I don't think we want to set this the same as the heartbeat timeout as that would affect all client connections, especially in the case where you might set your timeout to be zero.

aanm replied:
Oh right!

aanm commented on pkg/k8s/client.go:
```diff
@@ -105,11 +108,50 @@ func CreateConfig() (*rest.Config, error) {
+
+		select {
+		case err := <-done:
+			if err != nil {
```
In case the error is type of timeout (504 and / or 408?), we won't execute `closeAllConns`, is this expected?

tom-hadlaw-hs replied:
Yeah that makes sense, I changed it such that if there is a status error and either one of those status-codes, it'll restart the connection as well. Perhaps we'd want to add `StatusBadGateway` to that list?

aanm replied:
Yeah I guess we can add that one as well.

tom-hadlaw-hs committed:
Add heartbeat that can kill client connections

Adding a periodic heartbeat to check for stale connections to the
apiserver. If a heartbeat request times out then client-go's conntrack
dialer is used to close and reestablish all open connections.

Fixes: #9792

Signed-off-by: Tom Hadlaw <thomas.hadlaw@hootsuite.com>
